File: 543. Diameter of Binary Tree/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function(root) {
  let res = 0;
  getDepth(root);
  return res;

  function getDepth(node) {
    if (!node) return 0;
    const left = getDepth(node.left);
    const right = getDepth(node.right);
    res = Math.max(res, left + right);
    return Math.max(left, right) + 1;
  }
};

File: 543. Diameter of Binary Tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func diameterOfBinaryTree(root *TreeNode) int {
	maxLen := make([]int, 1)
	helper(root, maxLen)
	return maxLen[0]
}

func helper(root *TreeNode, maxLen []int) int {
	if root == nil {
		return 0
	}

	left := helper(root.Left, maxLen)
	right := helper(root.Right, maxLen)
	maxLen[0] = max(maxLen[0], left+right)
	return max(left, right) + 1
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}